consensus/fork_choice/src/fork_choice.rs

```diff
@@ -679,29 +677,81 @@ where
             self.fc_store.set_proposer_boost_root(block_root);
         }
 
-        // Update justified checkpoint.
-        if state.current_justified_checkpoint().epoch > self.fc_store.justified_checkpoint().epoch {
-            if state.current_justified_checkpoint().epoch
-                > self.fc_store.best_justified_checkpoint().epoch
+        // Update store with checkpoints if necessary
+        self.update_checkpoints(
+            state.current_justified_checkpoint(),
+            state.finalized_checkpoint(),
+            state.slot(),
+            current_slot,
+            spec,
+        )?;
+
+        // Update unrealized justified/finalized checkpoints.
+        let (unrealized_justified_checkpoint, unrealized_finalized_checkpoint) = if count_unrealized
+            .is_true()
+        {
+            let justification_and_finalization_state = match block {
+                BeaconBlockRef::Merge(_) | BeaconBlockRef::Altair(_) => {
+                    let participation_cache =
+                        per_epoch_processing::altair::ParticipationCache::new(state, spec)
+                            .map_err(Error::ParticipationCacheBuild)?;
+                    per_epoch_processing::altair::process_justification_and_finalization(
+                        state,
+                        &participation_cache,
+                    )?
+                }
+                BeaconBlockRef::Base(_) => {
+                    let mut validator_statuses =
+                        per_epoch_processing::base::ValidatorStatuses::new(state, spec)
+                            .map_err(Error::ValidatorStatuses)?;
+                    validator_statuses
+                        .process_attestations(state)
+                        .map_err(Error::ValidatorStatuses)?;
+                    per_epoch_processing::base::process_justification_and_finalization(
+                        state,
+                        &validator_statuses.total_balances,
+                        spec,
+                    )?
+                }
+            };
+
+            let unrealized_justified_checkpoint =
+                justification_and_finalization_state.current_justified_checkpoint();
+            let unrealized_finalized_checkpoint =
+                justification_and_finalization_state.finalized_checkpoint();
+
+            // Update best known unrealized justified & finalized checkpoints
+            if unrealized_justified_checkpoint.epoch
+                > self.fc_store.unrealized_justified_checkpoint().epoch
             {
                 self.fc_store
-                    .set_best_justified_checkpoint(state.current_justified_checkpoint());
+                    .set_unrealized_justified_checkpoint(unrealized_justified_checkpoint);
```

Comment:
I need to think a little bit more about whether or not we need to check here to ensure that the UJC is a descendant of the RJC.

Reply:
I'm confident that we don't need to consider this because realization happens in the first slot of an epoch which is always within `SAFE_SLOTS_TO_UPDATE_JUSTIFIED`.